### PR TITLE
Fix the route line updated ahead of user location in short straight-line route step.

### PR DESCRIPTION
**Pull request by ShanMa1991**
### Description
This Pr is to fix the issue where the route line endpoint is ahead of user location indicator on short straight-line route step when `routeLineTracksTraversal` enabled.

### Implementation
- [x] Fix the calculation of the `allRemainingPoints` when update the `navigationMapView.routeRemainingDistancesIndex` with `RouteProgress`, by avoiding the upcoming route point jumping to the middle of next step when the user is on the short straight-line route step. 
- [x] Add tests for the update. 

### Screenshots or Gifs
In the Example App, choose a route that contains the short straight-line route steps with waypoints below. And choose to start basic navigation with `routeLineTracksTraversal` enabled.
```
waypoints = [
.init(coordinate: .init(latitude: 55.026291, longitude: 30.799548)),
.init(coordinate: .init(latitude: 55.026720, longitude: 30.798226))
]
```
fix

### CODE REVIEW

**Comment by Udumft** [comment]
Could this also fix route line flickering shown in [this comment](https://github.com/mapbox-collab/ncs-collab/issues/450#issuecomment-1175820764)?

**Comment by ShanMa1991** [comment]
> Could this also fix route line flickering shown in [this comment](https://github.com/mapbox-collab/ncs-collab/issues/450#issuecomment-1175820764)?

No, I don't think so. The route lien flickering with puck move back and forth is caused by https://github.com/mapbox/mapbox-navigation-ios/blob/c8dc5988870d8bf1d1ec6d800544365c33b35787/Sources/MapboxNavigation/NavigationMapView.swift#L1842-L1855 
Because we asked the route line to sync with the user puck. When the user puck is receiving back and forth location data, it will trigger the route line offset re-calculating to sync with the puck, and cause the flickering issue. The root cause is the location data moving back and forth that passed to the user puck.

File: CHANGELOG.md [hunk 1]
```diff
@@ -56,6 +57,7 @@
 * Added the `layerPosition` parameter to the `NavigationMapView.show(_:layerPosition:legIndex:)` method for controlling the position of the main route layer while presenting routes. ([#3897](https://github.com/mapbox/mapbox-navigation-ios/pull/3897))
 * Fixed an issue where camera kept receiving updates even after stopping `MapboxNavigationService`. ([#3928](https://github.com/mapbox/mapbox-navigation-ios/pull/3928))
 * Added `NavigationViewController.showsContinuousAlternatives` and `CarPlayNavigationViewController.showsContinuousAlternatives` flags to toggle displaying alternative route's lines during navigation session. Use `NavigationMapView.show(continuousAlternatives:)` and `NavigationMapView.removeContinuousAlternativesRoutes()` methods for fine control over displayed routes. ([#3850](https://github.com/mapbox/mapbox-navigation-ios/pull/3850))
+* Each alternative route shown during turn-by-turn navigation is now annotated with the amount of time it saves or adds to the trip. To hide these annotations, set the `NavigationMapView.showsRelativeDurationOnContinuousAlternativeRoutes` property to `false`. ([#3956](https://github.com/mapbox/mapbox-navigation-ios/pull/3956))
```
Comment (MaximAlien) [inline CHANGELOG.md:60]: Looks like this changelog entry is unrelated to this PR.

File: Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift [hunk 2]
```diff
@@ -92,18 +92,19 @@ extension NavigationMapView {
          Find the count of remaining points in the current step.
          */
         let lineString = currentStepProgress.step.shape ?? LineString([])
-        if currentStepProgress.distanceTraveled <= 0 {
-            allRemainingPoints += currentLegSteps[currentLegProgress.stepIndex].dropLast().count
+        // If user hasn't arrived at current step. All the coordinates will be included to the remaiining points.
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/NavigationMapView+VanishingRouteLine.swift:95]:
```suggestion
        // If user hasn't arrived at current step. All the coordinates will be included to the remaining points.
```

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift [hunk 3]
```diff
@@ -6,6 +6,7 @@ import MapboxMaps
 @testable import MapboxCoreNavigation
 
 class VanishingRouteLineTests: TestCase {
+    let errorAllowed = 1e-6
```
Comment (MaximAlien) [inline Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:9]: This property reads like boolean one. What about `accuracyThreshold`?

File: Tests/MapboxNavigationTests/VanishingRouteLineTests.swift [hunk 4]
```diff
@@ -107,36 +128,78 @@ class VanishingRouteLineTests: TestCase {
...
+        XCTAssertEqual(routeProgress.currentLegProgress.currentStepProgress.distanceTraveled, 0.0)
         
-        // Because mapbox-directions-swift parses the route with one more duplicate coordinate in the last step of each route leg.
-        // The one leg route has one more coordinate compared with Android.
-        XCTAssertEqual(navigationMapView.routeRemainingDistancesIndex, 7)
+        var traveledPoints = 0
+        traveledPoints += currentLegPoints[0].count
+        traveledPoints += currentLegPoints[1].count
+        // Because user has arrived at the first coordinate of current step with distanceTraveled on this step as 0.0.
+        traveledPoints += 1
```
Comment (MaximAlien) [inline Tests/MapboxNavigationTests/VanishingRouteLineTests.swift:190]: Why not make it a one liner constant?

Comment (ShanMa1991) [reply Tests/MapboxNavigationTests/VanishingRouteLineTests.swift]: Thanks for the suggestion, the constant is applied instead of the steps calculation.

**Review by MaximAlien — APPROVED** [review]
